demos/coreMQTT_Agent/simple_sub_pub_demo.c

```diff
@@ -0,0 +1,560 @@
+                        pxSubscribeArgs->pSubscribeInfo->topicFilterLength,
+                        pxSubscribeArgs->pSubscribeInfo->pTopicFilter ) );
+        }
+    }
+
+    xTaskNotify( pxCommandContext->xTaskToNotify,
+                 ( uint32_t ) ( pxReturnInfo->returnCode ),
+                 eSetValueWithOverwrite );
+}
+
+/*-----------------------------------------------------------*/
+
+static BaseType_t prvWaitForCommandAcknowledgment( uint32_t * pulNotifiedValue )
+{
+    BaseType_t xReturn;
+
+    /* Wait for this task to get notified, passing out the value it gets
+     * notified with. */
+    xReturn = xTaskNotifyWait( 0,
+                               0,
+                               pulNotifiedValue,
+                               pdMS_TO_TICKS( mqttexampleMS_TO_WAIT_FOR_NOTIFICATION ) );
+    return xReturn;
+}
+
+/*-----------------------------------------------------------*/
+
+static void prvIncomingPublishCallback( void * pvIncomingPublishCallbackContext,
+                                        MQTTPublishInfo_t * pxPublishInfo )
+{
+    static char cTerminatedString[ mqttexampleSTRING_BUFFER_LENGTH ];
+
+    ( void ) pvIncomingPublishCallbackContext;
+
+    /* Create a message that contains the incoming MQTT payload to the logger,
+     * terminating the string first. */
+    if( pxPublishInfo->payloadLength < mqttexampleSTRING_BUFFER_LENGTH )
+    {
+        memcpy( ( void * ) cTerminatedString, pxPublishInfo->pPayload, pxPublishInfo->payloadLength );
+        cTerminatedString[ pxPublishInfo->payloadLength ] = 0x00;
+    }
+    else
+    {
+        memcpy( ( void * ) cTerminatedString, pxPublishInfo->pPayload, mqttexampleSTRING_BUFFER_LENGTH );
+        cTerminatedString[ mqttexampleSTRING_BUFFER_LENGTH - 1 ] = 0x00;
+    }
+
+    LogInfo( ( "Received incoming publish message %s", cTerminatedString ) );
+}
+
+/*-----------------------------------------------------------*/
+
+static bool prvSubscribeToTopic( MQTTQoS_t xQoS,
+                                 char * pcTopicFilter )
+{
+    MQTTStatus_t xCommandAdded;
+    BaseType_t xCommandAcknowledged = pdFALSE;
+    uint32_t ulSubscribeMessageID;
+    MQTTAgentSubscribeArgs_t xSubscribeArgs;
+    MQTTSubscribeInfo_t xSubscribeInfo;
+    static int32_t ulNextSubscribeMessageID = 0;
+    MQTTAgentCommandContext_t xApplicationDefinedContext = { 0 };
+    MQTTAgentCommandInfo_t xCommandParams = { 0 };
+
+    /* Create a unique number of the subscribe that is about to be sent.  The number
+     * is used as the command context and is sent back to this task as a notification
+     * in the callback that executed upon receipt of the subscription acknowledgment.
+     * That way this task can match an acknowledgment to a subscription. */
+    xTaskNotifyStateClear( NULL );
+    taskENTER_CRITICAL();
+    {
+        ulNextSubscribeMessageID++;
+        ulSubscribeMessageID = ulNextSubscribeMessageID;
+    }
+    taskEXIT_CRITICAL();
```

Comment:
Comment about why this is in the Critical section will be valuable

Reply:
There's already a comment block above this that explains that each subscribe is given a unique identifier for matching purposes

Reply:
The comment does not explicitly mention that the critical section is used for protecting access to the global variable,